generic/sosreport.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2017 IBM
+# Author: Pavithra <pavrampu@linux.vnet.ibm.com>
+
+import os
+import tempfile
+import shutil
+from avocado import Test
+from avocado import main
+from avocado.utils import process
+from avocado.utils import distro
+from avocado.utils.software_manager import SoftwareManager
+
+
+class sosreport_test(Test):
+
+    is_fail = 0
+
+    def run_cmd(self, cmd):
+        self.log.info("executing ============== %s =================" % cmd)
+        if process.system(cmd, ignore_status=True, sudo=True):
+            self.log.info("%s command failed" % cmd)
+            self.is_fail += 1
+        return
+
+    def run_cmd_out(self, cmd):
+        return process.system_output(cmd, shell=True, ignore_status=True, sudo=True)
+
+    def setUp(self):
+        dist = distro.detect()
+        sm = SoftwareManager()
+        if not process.system("sosreport", ignore_status=True, sudo=True):
+            self.log.info("sosreport is installed")
+        elif dist.name == 'Ubuntu':
+            if not sm.install("sosreport"):
+                self.skip("Fail to install sosreport required for this test.")
+        elif dist.name == 'redhat':
+            if not sm.install("sos"):
+                self.skip("Fail to install sos package required for this test.")
+        else:
+            self.skip("sosreport is not supported on this distro "
+                      "or package is not installed")
```

Comment:
why not it can be done in this way 
```

sos_pkg=""
if dist.name == 'Ubuntu':
  sos_pkg='sosreport'
elif dist.name == 'redhat':
  sos_pkg='sos'
else:
  self.skip("sosreport is not supported on this distro ")

if sos_pkg and not smm.check_installed(sos_pkg) and not smm.install(sos_pkg):
   self.skip("Package %s is missing and could not be installed" % pkg)
```